bindings/c/src/common.rs

```diff
@@ -851,23 +851,49 @@ impl node_builder {
     }
 }
 
-#[repr(C)]
 pub struct tree {
-    pub root: node_id,
+    _private: [u8; 0],
+}
+
+impl CastPtr for tree {
+    type RustType = Tree;
 }
 
+impl BoxCastPtr for tree {}
+
 impl tree {
     #[no_mangle]
-    pub extern "C" fn accesskit_tree_new(root: node_id) -> tree {
-        tree { root }
+    pub extern "C" fn accesskit_tree_new(root: node_id) -> *mut tree {
+        let t = Tree::new(root.into());
+        BoxCastPtr::to_mut_ptr(t)
     }
-}
 
-impl From<tree> for Tree {
-    fn from(tree: tree) -> Self {
-        Self {
-            root: tree.root.into(),
-        }
+    #[no_mangle]
+    pub extern "C" fn accesskit_tree_free(t: *mut tree) {
+        drop(box_from_ptr(t));
+    }
+
+    #[no_mangle]
+    pub extern "C" fn accesskit_tree_set_app_name(t: *mut tree, app_name: *const c_char) {
+        let t = mut_from_ptr(t);
+        t.app_name = Some(String::from(
+            unsafe { CStr::from_ptr(app_name) }.to_string_lossy(),
+        ));
+    }
+
+    #[no_mangle]
+    pub extern "C" fn accesskit_tree_set_toolkit_info(
```

Comment:
I understand why you chose to have one function here, I'm in favor of keeping the C bindings as close to the Rust schema as possible though.